File: medical-app-frontend/src/app/services/consultation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { Consultation } from '../models/consultation.model';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class ConsultationService {
  private apiUrl = `${environment.apiUrl}/consultation`;

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) {}

  private getHeaders(): Observable<HttpHeaders> {
    return this.authService.getAuthToken().pipe(
      map(token => {
        if (!token) throw new Error('No authentication token available');
        return new HttpHeaders({
          'Authorization': `Bearer ${token}`
        });
      })
    );
  }

  getConsultationByAppointment(appointmentId: string): Observable<Consultation | null> {
    return this.getHeaders().pipe(
      switchMap(headers =>
        this.http.get<any>(`${this.apiUrl}/consultations/appointment/${appointmentId}`, { headers }).pipe(
          map(consultation => ({
            ...consultation,
            id: consultation._id
          })),
          catchError(error => {
            if (error.status === 404) {
              return throwError(() => null);
            } else if (error.status === 403) {
              console.log('Unauthorized: Not your consultation');
              return throwError(() => new Error('Unauthorized: You do not have permission to view this consultation'));
            }
            console.error('Error fetching consultation by appointment:', error);
            return throwError(() =>
              new Error(error.error?.error || 'Failed to fetch consultation')
            );
          })
        )
      )
    );
  }

  getConsultationByAppointmentPatientVersion(appointmentId: string): Observable<Consultation | null> {
    return this.getHeaders().pipe(
      switchMap(headers =>
        this.http.get<any>(`${this.apiUrl}/patient/consultation/appointment/${appointmentId}`, { headers }).pipe(
          map(consultation => ({
            ...consultation,
            id: consultation._id
          })),
          catchError(error => {
            if (error.status === 404) {
              return throwError(() => null);
            } else if (error.status === 403) {
              console.log('Unauthorized: Not your consultation');
              return throwError(() => new Error('Unauthorized: You do not have permission to view this consultation'));
            }
            console.error('Error fetching consultation by appointment:', error);
            return throwError(() =>
              new Error(error.error?.error || 'Failed to fetch consultation')
            );
          })
        )
      )
    );
  }

  getPatientConsultations(): Observable<Consultation[]> {
    return this.getHeaders().pipe(
      switchMap(headers =>
        this.http.get<Consultation[]>(`${this.apiUrl}/patient/consultations`, { headers })
      ),
      map(list =>
        list.map(c => ({
          ...c,
          id: c.id
        }))
      ),
      catchError(error => {
        console.error('Error fetching patient consultations:', error);
        return throwError(() =>
          new Error(error.error?.error || 'Failed to fetch patient consultations')
        );
      })
    );
  }

  deleteConsultation(consultationId: string): Observable<void> {
    if (!consultationId) {
      console.error('Missing consultation ID!');
      return throwError(() => new Error('Missing consultation ID!'));
    }
    return this.getHeaders().pipe(
      switchMap(headers =>
        this.http.delete<void>(`${this.apiUrl}/consultations/${consultationId}`, { headers })
      ),
      catchError(error => {
        console.error('Error deleting consultation:', error);
        return throwError(() =>
          new Error(error.error?.error || 'Failed to delete consultation')
        );
      })
    );
  }

  createConsultation(
    consultation: Partial<Consultation>,
    files: File[] = []
  ): Observable<Consultation> {
    return this.getHeaders().pipe(
      switchMap(headers => {
        const formData = new FormData();
        Object.entries(consultation).forEach(([key, value]) => {
          if (value !== undefined && value !== null) {
            const fieldValue = typeof value === 'object' ? JSON.stringify(value) : String(value);
            formData.append(key, fieldValue);
          }
        });
        files.forEach(file => {
          formData.append('files', file, file.name);
        });
        return this.http.post<Consultation>(`${this.apiUrl}/consultations`, formData, { headers });
      }),
      catchError(error => {
        console.error('Error creating consultation:', error);
        return throwError(() =>
          new Error(error.error?.error || 'Failed to create consultation')
        );
      })
    );
  }


  updateConsultation(id: string, updates: Partial<Consultation>): Observable<void> {
    return this.getHeaders().pipe(
      switchMap((headers: HttpHeaders) =>
        this.http.put<void>(`${this.apiUrl}/consultations/${id}`, updates, { headers })
      ),
      catchError(error => {
        console.error('Error updating consultation:', error);
        return throwError(() => new Error(error.error?.error || 'Failed to update consultation'));
      })
    );
  }

  downloadDocument(consultationId: string, filename: string): Observable<Blob> {
    return this.getHeaders().pipe(
      switchMap((headers: HttpHeaders) =>
        this.http.get(`${this.apiUrl}/consultations/${consultationId}/documents/${filename}`, {
          headers,
          responseType: 'blob'
        })
      ),
      catchError(error => {
        console.error('Error downloading document:', error);
        return throwError(() => new Error(error.error?.error || 'Failed to download document'));
      })
    );
  }
}

File: medical-app-backend/services/consultation_service.py
from models.Consultation import create_consultation, find_consultation_by_id, find_consultations_by_doctor, find_consultations_by_patient, update_consultation, delete_consultation
from models.User import find_user_by_id

def validate_consultation_data(data, doctor_id):
    required_fields = ['appointmentId', 'patientId', 'diagnosis', 'prescription']
    for field in required_fields:
        if field not in data or not data[field]:
            return False, f"Missing or invalid {field}"
    if data['doctorId'] != doctor_id:
        return False, "Unauthorized: Doctor ID mismatch"
    # Verify patient and doctor exist
    if not find_user_by_id(data['patientId']):
        return False, "Patient not found"
    if not find_user_by_id(data['doctorId']):
        return False, "Doctor not found"
    return True, None

def create_new_consultation(consultation_data, doctor_id, uploaded_files=None):
    print(f"Creating new consultation with data: {consultation_data}") # Debug log
    is_valid, error = validate_consultation_data(consultation_data, doctor_id)
    if not is_valid:
        return None, error
    consultation = create_consultation(consultation_data, uploaded_files)
    if not consultation:
        return None, "Failed to create consultation"
    return consultation, None

def get_consultation(consultation_id, doctor_id):
    print(f"Fetching consultation ID: {consultation_id}") # Debug log
    consultation = find_consultation_by_id(consultation_id)
    if not consultation:
        return None, "Consultation not found"
    if consultation['doctorId'] != doctor_id:
        return None, "Unauthorized: Not your consultation"
    return consultation, None

def get_doctor_consultations(doctor_id):
    print(f"Fetching consultations for doctor: {doctor_id}") # Debug log
    consultations = find_consultations_by_doctor(doctor_id)
    return consultations, None

def get_patient_consultations(patient_id, requesting_user_id):
    print(f"Fetching consultations for patient: {patient_id}") # Debug log
    if patient_id != requesting_user_id:
        return None, "Unauthorized: You can only view your own consultations"
    consultations = find_consultations_by_patient(patient_id)
    return consultations, None

def update_existing_consultation(consultation_id, updates, doctor_id):
    print(f"Updating consultation ID: {consultation_id} with updates: {updates}") # Debug log
    consultation = find_consultation_by_id(consultation_id)
    if not consultation:
        return False, "Consultation not found"
    if consultation['doctorId'] != doctor_id:
        return False, "Unauthorized: Not your consultation"
    success = update_consultation(consultation_id, updates)
    if not success:
        return False, "Failed to update consultation"
    return True, None

def delete_existing_consultation(consultation_id, doctor_id):
    print(f"Deleting consultation ID: {consultation_id}") # Debug log
    consultation = find_consultation_by_id(consultation_id)
    if not consultation:
        return False, "Consultation not found"
    if consultation['doctorId'] != doctor_id:
        return False, "Unauthorized: Not your consultation"
    success = delete_consultation(consultation_id)
    if not success:
        return False, "Failed to delete consultation"
    return True, None